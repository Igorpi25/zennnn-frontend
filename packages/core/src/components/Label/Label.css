.label {
  @apply flex items-center relative h-5 my-2;
}

.label__wrapper {
  @apply absolute inset-0 inline-flex;
}

.label__content__wrapper {
  @apply leading-5 max-w-full;
}

.label__content {
  @apply text-base text-gray-300 dark:text-gray-100;
  @apply relative w-full truncate;
}
.label--show-wrap .label__content {
  @apply pointer-events-none;
}
