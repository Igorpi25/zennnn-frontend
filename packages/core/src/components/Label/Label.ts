import {
  h,
  ref,
  defineComponent,
} from 'vue'

import uid from '../../utils/uid'

export default defineComponent({
  name: 'Label',

  props: {
    id: String,
    for: String,
    value: String,
    showWrap: Boolean,
  },

  setup (props, { slots }) {
    const id: string = props.id || uid('label-')
    const rootElement = ref<HTMLElement>()
    const contentWrapperElement = ref<HTMLElement>()
    const contentElement = ref<HTMLElement>()
    const closestBackgroundColor = ref<string>('transparent')
    const hasOverflow = ref<boolean>(false)
    let timeout: any = null

    const getClosestBackgroundColor = () => {
      let el = rootElement.value
      while (el?.parentNode) {
        const { backgroundColor } = getComputedStyle(el)
        if (backgroundColor !== 'transparent' && backgroundColor !== 'rgba(0, 0, 0, 0)') {
          closestBackgroundColor.value = backgroundColor
          break
        }
        el = el.parentNode as HTMLElement
      }
    }

    const onPointerenter = () => {
      timeout = setTimeout(() => {
        const clientWidth = contentWrapperElement.value?.clientWidth || 0
        const scrollWidth = contentElement.value?.scrollWidth || 0
        hasOverflow.value = scrollWidth > clientWidth
        if (hasOverflow.value) {
          getClosestBackgroundColor()
          contentElement.value!.style.backgroundColor = closestBackgroundColor.value
          contentElement.value!.style.overflow = 'visible'
          contentElement.value!.style.whiteSpace = 'normal'
          contentElement.value!.style.zIndex = '1'
        }
      }, 120)
    }
    const onPointerleave = () => {
      if (timeout) {
        clearTimeout(timeout)
        timeout = null
      }
      if (hasOverflow.value) {
        contentElement.value!.style.backgroundColor = 'unset'
        contentElement.value!.style.overflow = 'hidden'
        contentElement.value!.style.whiteSpace = 'nowrap'
        contentElement.value!.style.zIndex = 'unset'
      }
    }

    const genContentWrapper = () => {
      const data = {
        ref: contentWrapperElement,
        class: 'label__content__wrapper',
      } as any
      if (props.showWrap) {
        data.onPointerenter = onPointerenter
        data.onPointerleave = onPointerleave
      }
      return h('div', data, genContent())
    }

    const genAppendSlot = () => {
      return slots.append?.()
    }

    const genContent = () => {
      return h('div', {
        ref: contentElement,
        class: 'label__content',
      }, getContent())
    }

    const getContent = () => {
      return slots.default?.() ?? props.value
    }

    return () => {
      return h('label', {
        ref: rootElement,
        id: id,
        for: props.for,
        'aria-hidden': !props.for,
        class: {
          label: true,
          'label--show-wrap': props.showWrap,
        },
      }, h('div', {
        class: 'label__wrapper',
      }, [genContentWrapper(), genAppendSlot()]))
    }
  },
})
