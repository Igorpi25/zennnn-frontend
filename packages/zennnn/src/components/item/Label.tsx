import { defineComponent, computed, PropType, CSSProperties } from 'vue'
import { convertToUnit } from 'vue-supp'
import { LabelTypes } from '@/components/item/Card'

export interface ItemLabelProps {
  preset?: LabelTypes
  background?: string
  count?: number | string
  maxWidth?: number | string
}

export default defineComponent({
  props: {
    preset: null as any as PropType<LabelTypes>,
    background: String,
    count: [Number, String] as PropType<number | string>,
    maxWidth: [Number, String] as PropType<number | string>,
  },

  setup(props, { slots }) {
    const background = computed(() => {
      if (!props.preset) return props.background
      switch (props.preset) {
        case LabelTypes.New:
          return '#34B0B0'
        case LabelTypes.Sale:
          return '#0476FF'
        case LabelTypes.Models:
          return '#A26014'
        case LabelTypes.Colors:
          return 'linear-gradient(270deg, #C51D1D 0%, #146491 50%, #0F761A 100%)'
        case LabelTypes.Sizes:
          return '#4634B0'
        default:
          return props.background
      }
    })

    const text = computed(() => {
      if (!props.preset) return undefined
      switch (props.preset) {
        case LabelTypes.New:
          return 'New'
        case LabelTypes.Sale:
          return 'Sale'
        case LabelTypes.Models:
          return 'Models'
        case LabelTypes.Colors:
          return 'Colors'
        case LabelTypes.Sizes:
          return 'Sizes'
        default:
          return undefined
      }
    })

    const styles = computed(() => {
      const s = {} as CSSProperties
      if (background.value) {
        s.background = background.value
      }
      if (props.maxWidth) {
        s.maxWidth = convertToUnit(props.maxWidth)
      }
      return s
    })

    return () => (
      <div
        class={{
          'h-4.5 inline-flex items-center text-xs text-white rounded-10 space-x-1 truncate px-1':
            true,
          'pr-0.5': props.count,
        }}
        style={styles.value}
      >
        <span>{slots.default ? slots.default() : text.value}</span>
        {props.count && (
          <span
            class="inline-flex items-center justify-center rounded-full bg-light-gray-400 text-gray-900"
            style={{ height: '14px', minWidth: '14px' }}
          >
            {props.count}
          </span>
        )}
      </div>
    )
  },
})
