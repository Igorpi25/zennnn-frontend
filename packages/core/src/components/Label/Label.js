import {
  h,
  ref,
} from 'vue'

import uid from '../../utils/uid'

import './Label.css'

export default {
  name: 'Label',

  props: {
    id: String,
    for: String,
    value: String,
    showWrap: Boolean,
  },

  setup (props, { slots }) {
    const id = props.id || uid('label-')
    const rootElement = ref(null)
    const contentWrapperElement = ref(null)
    const contentElement = ref(null)
    const closestBackgroundColor = ref('transparent')
    const hasOverflow = ref(false)
    let timeout = null

    const getClosestBackgroundColor = () => {
      let el = rootElement.value
      while (el.parentNode) {
        const { backgroundColor } = getComputedStyle(el)
        if (backgroundColor !== 'transparent' && backgroundColor !== 'rgba(0, 0, 0, 0)') {
          closestBackgroundColor.value = backgroundColor
          break
        }
        el = el.parentNode
      }
    }

    const onPointerenter = () => {
      timeout = setTimeout(() => {
        const clientWidth = contentWrapperElement.value && contentWrapperElement.value.clientWidth
        const scrollWidth = contentElement.value && contentElement.value.scrollWidth
        hasOverflow.value = scrollWidth > clientWidth
        if (hasOverflow.value) {
          getClosestBackgroundColor()
          contentElement.value.style.backgroundColor = closestBackgroundColor.value
          contentElement.value.style.overflow = 'visible'
          contentElement.value.style.whiteSpace = 'normal'
          contentElement.value.style.zIndex = 1
        }
      }, 120)
    }
    const onPointerleave = () => {
      if (timeout) {
        clearTimeout(timeout)
        timeout = null
      }
      if (hasOverflow.value) {
        contentElement.value.style.backgroundColor = 'unset'
        contentElement.value.style.overflow = 'hidden'
        contentElement.value.style.whiteSpace = 'nowrap'
        contentElement.value.style.zIndex = 'unset'
      }
    }

    const genContentWrapper = () => {
      const data = {
        ref: contentWrapperElement,
        class: 'label__content__wrapper',
      }
      if (props.showWrap) {
        data.onPointerenter = onPointerenter
        data.onPointerleave = onPointerleave
      }
      return h('div', data, genContent())
    }

    const genAppendSlot = () => {
      return slots.append && slots.append()
    }

    const genContent = () => {
      return h('div', {
        ref: contentElement,
        class: 'label__content',
      }, getContent())
    }

    const getContent = () => {
      return slots.default ? slots.default() : props.value
    }

    return {
      rootElement,
      id,
      closestBackgroundColor,
      genContentWrapper,
      genAppendSlot,
    }
  },

  render () {
    return h('label', {
      ref: 'rootElement',
      id: this.id,
      for: this.for,
      'aria-hidden': !this.for,
      class: {
        label: true,
        'label--show-wrap': this.showWrap,
      },
    }, h('div', {
      class: 'label__wrapper',
    }, [this.genContentWrapper(), this.genAppendSlot()]))
  },
}
