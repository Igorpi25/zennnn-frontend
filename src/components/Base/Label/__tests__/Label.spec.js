import { mount } from '@vue/test-utils'
import Label from '../Label'

jest.mock('../../../../utils/uid')

describe('Label.js', () => {
  let mountFunction

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Label, {
        ...options,
      })
    }
  })

  // Render

  it('should render component with default slot and match snapshot', () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with id and value prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        id: 'label-id',
        value: 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with for prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        for: 'input-id',
        value: 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with append slot and match snapshot', () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
        append: () => 'Append slot',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // TODO: need use attachTo for set max-width to parent
  it.todo('should show wrapped content on pointer enter')

  // TODO: need use attachTo for set background to parent
  it.todo('should set correct background-color on pointer enter')
})
