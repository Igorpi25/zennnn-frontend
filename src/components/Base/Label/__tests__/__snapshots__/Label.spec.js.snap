// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label.js should render component with append slot and match snapshot 1`] = `
<label id="label-1" aria-hidden="true" class="label">
  <div class="label__wrapper">
    <div class="label__content__wrapper">
      <div class="label__content">Label</div>
    </div>Append slot
  </div>
</label>
`;

exports[`Label.js should render component with default slot and match snapshot 1`] = `
<label id="label-1" aria-hidden="true" class="label">
  <div class="label__wrapper">
    <div class="label__content__wrapper">
      <div class="label__content">Label</div>
    </div>
    <!---->
  </div>
</label>
`;

exports[`Label.js should render component with for prop and match snapshot 1`] = `
<label id="label-1" for="input-id" aria-hidden="false" class="label">
  <div class="label__wrapper">
    <div class="label__content__wrapper">
      <div class="label__content">Label</div>
    </div>
    <!---->
  </div>
</label>
`;

exports[`Label.js should render component with id and value prop and match snapshot 1`] = `
<label id="label-id" aria-hidden="true" class="label">
  <div class="label__wrapper">
    <div class="label__content__wrapper">
      <div class="label__content">Label</div>
    </div>
    <!---->
  </div>
</label>
`;
